### DISCUSSION TITLE
Is it possible to install ToaruOS on a hard drive?

### DISCUSSION THREAD

**Discussion by mxlgv** [klange/toaruos#274] [Q&A] [answered]
I apologize right away. I may not have learned much about the system, but I would like to be able to save ToaruOS files and settings on the hard drive. Apart from this, I don't know if there is a "connection with the outside world" from the virtual machine. Now I use a small http server on the host system and the fetch utility to transfer files to ToaruOS.

**Comment by klange** [comment, marked as answer]
ToaruOS does have drivers for ATA hard disks, as well as an incomplete write-capable EXT2 filesystem implementation, but due to neglect they are both known to be buggy. Through a manual process, ToaruOS can be installed to a hard disk alongside Grub as a bootloader, packages can generally be installed, and with some luck files will persist between reboots, until something inevitably goes wrong. No automated installer has been developed, and the in-house bootloader does not support loading files from EXT2 filesystems.

Instead of doing a full install, you can create a hard disk image, format it either directly as EXT2 or using an MBR partition table with an EXT2 partition, and provide that to your VM. You can then install the `ata` and `dospart` modules, find your partition in `/dev`, and mount it somewhere. If you're willing to do a custom build, you can also modify the startup scripts on the CD image's ramdisk to do this automatically.

> Now I use a small http server on the host system and the fetch utility to transfer files to ToaruOS.

I do the same thing =)

There's also a tool called `upload.krk` that will toss files over TCP sockets that you can use with `nc` on a host to get files in the other direction. I usually use it to get native screenshots out of a VM:

```bash
# Host
$ nc -l 6024 > /tmp/test.tga && convert -auto-orient /tmp/test.tga /tmp/toaru_screen_$(date +%F_%H_%M_%S).png
# Toaru
$ upload.krk /tmp/screenshot.png 192.168.1.1:6024 # or whatever your host's local IP is
```

**Reply by mxlgv** [reply]
Thanks for the answer. Can I use fat32 instead of EXT2?

**Reply by klange** [reply]
There's no fat32 driver, so no - unless you want to write one yourself.

#174 is an old tracking ticket for implementing fat32 and fat16.

**Reply by mxlgv** [reply]
It's clear. Let's see, file systems are still difficult for me.
